### ISSUE TITLE
Checkov hangs with CHECKOV_ENABLE_MODULES_FOREACH_HANDLING=True in a multi-workspace/multi-module project

### ISSUE THREAD

**Issue by ankon** [bridgecrewio/checkov#5047]
**Describe the issue**

Checkov since #5021 hangs in our CI (reproducible also locally). Setting `CHECKOV_ENABLE_MODULES_FOREACH_HANDLING` to `False` in the environment fixes the hang. 

Found true bisecting from 2.3.213 on to current master.

**Examples**

Cannot share/unknown (we're talking about 100 workspaces and a bunch of internal modules here!)

**Exception Trace**
Please share the trace for the exception and all relevant output by checkov.
To maximize the understanding, please run checkov with LOG_LEVEL set to debug
as follows:
```sh
LOG_LEVEL=DEBUG checkov ...
```

**Desktop (please complete the following information):**
 - OS: Linux, macOS
 - Checkov Version [e.g. 22]: `ab54127d7 Fri, 5 May 2023 03:17:34 +0000 (HEAD -> main, tag: 2.3.224, origin/main) platform(general): Catch None responses from BE (#5033) [GitHub Actions Bot]`

**Comment by ankon** [comment]
There is no exception trace, but here are the last words of checkov before going into 100% single-core CPU usage mode, where Ctrl-C is ineffective:

```
2023-05-05 10:40:29,162 [MainThread  ] [DEBUG]  Module load loop 0
2023-05-05 10:40:34,907 [MainThread  ] [INFO ]  Building graph from parsed module
2023-05-05 10:40:34,908 [MainThread  ] [INFO ]  Creating vertices
2023-05-05 10:40:35,345 [MainThread  ] [INFO ]  [TerraformLocalGraph] created 29836 vertices
2023-05-05 10:40:35,346 [MainThread  ] [INFO ]  Creating edges
2023-05-05 10:40:38,356 [MainThread  ] [INFO ]  [TerraformLocalGraph] created 46731 edges
2023-05-05 10:40:38,357 [MainThread  ] [INFO ]  [TerraformLocalGraph] start handling foreach
2023-05-05 10:40:40,381 [MainThread  ] [DEBUG]  evaluating 13050 edges
2023-05-05 10:40:40,522 [MainThread  ] [DEBUG]  got a substring with double underscore, which is not allowed. origin string: (elided)
2023-05-05 10:40:40,523 [MainThread  ] [DEBUG]  got a substring with double underscore, which is not allowed. origin string: (elided)
2023-05-05 10:40:42,021 [MainThread  ] [DEBUG]  evaluating 5089 edges
2023-05-05 10:40:42,547 [MainThread  ] [DEBUG]  evaluating 4958 edges
2023-05-05 10:40:43,901 [MainThread  ] [DEBUG]  evaluating 10558 edges
2023-05-05 10:40:45,381 [MainThread  ] [DEBUG]  evaluating 7183 edges
2023-05-05 10:40:46,130 [MainThread  ] [DEBUG]  evaluating 12129 edges
2023-05-05 10:40:46,980 [MainThread  ] [DEBUG]  evaluating 8845 edges
2023-05-05 10:40:47,601 [MainThread  ] [DEBUG]  evaluating 14268 edges
2023-05-05 10:40:48,264 [MainThread  ] [DEBUG]  evaluating 10061 edges
2023-05-05 10:40:48,891 [MainThread  ] [DEBUG]  evaluating 15316 edges
2023-05-05 10:40:49,630 [MainThread  ] [DEBUG]  evaluating 10607 edges
2023-05-05 10:40:50,233 [MainThread  ] [DEBUG]  evaluating 15555 edges
2023-05-05 10:40:50,936 [MainThread  ] [DEBUG]  evaluating 10688 edges
2023-05-05 10:40:51,546 [MainThread  ] [INFO ]  Reached too many edge duplications of 90% for 4 iterations. breaking.
```

**Comment by gruebel** [comment]
hey @ankon we tested this feature against our customer code base and didn't see such a behaviour yet. Since it is related to the `for_each` on module level feature, I assume you have a lot of `for_each/count` used on modules?
Without any example code we can't do much, but looking out, if others have the same issue with a smaller code base.

**Comment by ankon** [comment]
Yes, we're using `for_each` and `count` quite often, both in modules, as well as when invoking modules. 

I'm happy to try debugging steps here to get closer to the problem, but I cannot share the code base itself.

**Comment by dmnord** [comment]
I have a similar issue with running checkov. Our code in Terraform contains a huge number of modules with for_each and count.
In my case I face the issue that looks like the following:
![image](https://github.com/bridgecrewio/checkov/assets/12894352/79b4f04f-477a-4471-abdb-0846847a59f5)

The error happens when nested module is checked during the run. If this module is commented in the code then everything is going fine, and vise versa if this module is uncommented out in the code then the issue happens. This module is being called and used in different places across our Terraform code and only one single place in the code -- when a nested call is used -- produces  this error. Though, this module is always nested in different places within a code.

I tested this flag (CHECKOV_ENABLE_MODULES_FOREACH_HANDLING=true) setting it to false and suddenly I realized that the issue was solving  when this parameter was used.

@gruebel, can it be considered as a bug?
